subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/DefaultPrecompiledScriptPluginsSupport.kt

```diff
@@ -343,23 +386,15 @@ fun Project.exposeScriptsAsGradlePlugins(scriptPlugins: List<PrecompiledScriptPl
 
 
 private
-fun Project.collectScriptPlugins(): List<PrecompiledScriptPlugin> =
-    collectScriptPluginFiles().map(::PrecompiledScriptPlugin)
-
-
-private
-fun Project.collectScriptPluginFiles(): List<File> =
-    mutableListOf<File>().apply {
-        // TODO:configuration-cache undeclared build logic input
-        gradlePlugin.pluginSourceSet.allSource.matching {
-            it.include("**/*.gradle.kts")
-        }.visit {
-            if (!it.isDirectory) {
-                // TODO: preserve it.relativePath in PrecompiledScriptPlugin
-                add(it.file)
-            }
-        }
-    }
+fun Project.collectScriptPluginFiles(): Set<File> =
+    gradlePlugin.pluginSourceSet.allSource.matching {
+        it.include("**/*.gradle.kts")
+    }.filter {
+        !it.isDirectory
+    }.also {
+        // Declare build configuration input
+        fileCollectionObserved(it, "Kotlin DSL")
```

Comment:
:question: This is tricky. Here we explicitly says that observing this FC is a build logic input. Shouldn't this be detected instead?

I'm not saying we need to do it now but it looks like a valid use case we may want to support eventually.